test/pydmap_tests/test_dmapread.py

```diff
@@ -0,0 +1,325 @@
+        self.assertIsInstance(dm_records[0], collections.OrderedDict)
+        self.assertIsInstance(dm_records[2]['channel'], pydarn.DmapScalar)
+        self.assertIsInstance(dm_records[5]['ptab'], pydarn.DmapArray)
+        self.assertIsInstance(dm_records[9]['channel'].value, int)
+        self.assertIsInstance(dm_records[4]['ptab'].value, np.ndarray)
+        self.assertEqual(dm_records[0]['scan'].value, 1)
+        self.assertEqual(dm_records[7]['ltab'].dimension, 2)
+
+    def test_read_map(self):
+        """
+        Test reading records from map file.
+
+        Checks:
+            - returns correct data structures
+            - returns excpected values
+        """
+        file_path = map_file
+        dm = pydarn.DmapRead(file_path)
+        dm_records = dm.read_records()
+        self.assertIsInstance(dm_records, collections.deque)
+        self.assertIsInstance(dm_records[0], collections.OrderedDict)
+        self.assertIsInstance(dm_records[2]['IMF.flag'], pydarn.pydmap.datastructures.DmapScalar)
+        self.assertIsInstance(dm_records[3]['stid'], pydarn.DmapArray)
+        self.assertIsInstance(dm_records[8]['IMF.flag'].value, int)
+        self.assertIsInstance(dm_records[10]['stid'].value, np.ndarray)
+        self.assertEqual(dm_records[3]['stid'].dimension, 1)
+        self.assertEqual(dm_records[0]['stid'].shape[0], 14)  # this will be file dependent... future working test project.
+
+    # TODO: Again dependent on the file used :/
+    def test_integrity_check_corrupt_file1(self):
+        """
+        Test test_initial_data_integrity on a corrupt file
+
+        Expected bahaviour: raises pydmap expection
+        """
+        dmap = pydarn.DmapRead(corrupt_file1)
+        with self.assertRaises(pydarn.pydmap_exceptions.MismatchByteError):
+            dmap.test_initial_data_integrity()
+
+    def test_read_corrupt_file1(self):
+        """
+        Test read_records on a corrupt file
+
+        Expected bahaviour: raises pydmap expection
+        """
+        dmap = pydarn.DmapRead(corrupt_file1)
+        with self.assertRaises(pydarn.pydmap_exceptions.DmapDataTypeError):
+            dmap.read_records()
+
+    def test_integrity_check_corrupt_file2(self):
+        """
+        Test test_initial_data_integrity on a corrupt file
+
+        Expected bahaviour: raises pydmap expection
+        """
+        dmap = pydarn.DmapRead(corrupt_file2)
+        with self.assertRaises(pydarn.pydmap_exceptions.ZeroByteError):
+            dmap.test_initial_data_integrity()
+
+    def test_read_currupt_file2(self):
+        """
+        Test read_records on a corrupt file
+
+        Expected bahaviour: raises pydmap expection
+        """
+        dmap = pydarn.DmapRead(corrupt_file2)
+        with self.assertRaises(pydarn.pydmap_exceptions.ZeroByteError):
+            dmap.read_records()
+
+    # FIXME: have not gotten this test to work with the streaming. I believe it is an utf-8 problem
```

Comment:
Unfortunate. When I was testing, this was the single most valuable test.

Reply:
Why? 

I think the ability to read a DMAP file shows a lot of value and it to detect an error in bad files is just as valuable.

Reply:
Because scrambling bits randomly hit almost all the test cases in where you could have corruption. I was able to use this method for hitting close to 100% code coverage.

Reply:
This was important to me at the time since the realtime feed from prince george was crashing the backend every five minutes.

Reply:
I can see that thus the FIXME tag. 

Will review when I get to focus on streaming with DmapRead. 

Realtime is out of scope for this PR and the test code that needs to be fixed.